# Learn how to build your own Multi-Agent Fundamental Analysis Workflow with LlamaIndex


## Agents
By leveraging on AgentWorflow class from LlamaIndex, we'll create a multi-agent system composed of:

- A fundamental_agent: Responsible for collecting ratios for a given ticker using the FinanceToolkit package

- A profitability_agent: Responsible for gathering profitability ratios (ROA, ROE, Net Profit Margin, and Gross Margin) and analyzing their strengths or weaknesses compared to given thresholds. For example, if ROA > 5%, the firm is in a healthy range.

- A liquidity_agent: This agent collects liquidity ratios for a given ticker (Current Ratio, Quick Ratio, Debt-to-Equity Ratio, Interest Coverage Ratio) and comments on these ratios given a set of threshold values. For example if Current Ratio is between 1.5-3.0 the firm is in a healthy range.

- A supervisor_agent: This agent provides a final comment on the overall health of the firm based on the different comments coming from the different agents.

I've limited my agents to 2 specialized ones, you can add as much as you need.

## Agent's Breakdown:
Then for each agent:

- A clear description of the goal of the agent was provided

- A system prompt of the agent's instructions

- A Tool (you can also provide several tools) that helps gathering the data needed and compare it to the set of the pre-defined thresholds.

- The next agent to take the handoff.

## Tools
There are 4 tools, each agent has a tool. Each tool can act and modify the State of the multi-agent system.


## AgentWorkflow
Finally, the AgentWorkflow class puts all these together:

- You list the agents to use

- You specify the agent root

- You initiate your state

## LLMs:

I used GPT-4o and compare it to the latest model of OpenAI GPT-4.5-preview.
(I've used function calling because these agents supports natively tool calling).

## Key Takeaways:
- The workflow is easy to put in place
- The agent's breakdown is well-defined, from its description to the system prompt and the next agents for handoff.
- I called LLMs inside of the tools to provide insights on the ratios compared to the thresholds. Otherwise, I faced sometimes some hallucination...
- The final answer from GPT-4.5-preview is well-structured compared to GPT-4o, providing key ratios, values, comments, and overall insights.

In [ ]:
%pip install llama-index -q
!pip install financetoolkit -q

In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
from google.colab import userdata
FINANCIAL_MODELING_PREP_API_KEY = userdata.get('FINANCIAL_MODELING_PREP_API_KEY')
OPENAI_API_KEY = userdata.get('OPENAI_API_KEY')

from financetoolkit import Toolkit
import pandas as pd

from llama_index.llms.openai import OpenAI

llm = OpenAI(model="gpt-4o", api_key=OPENAI_API_KEY)


llm_45 = OpenAI(model="gpt-4.5-preview", api_key=OPENAI_API_KEY)


# LLM: GPT-4o

## Tools

In [ ]:
from llama_index.core.workflow import Context

async def get_fundamentals(ctx: Context, ticker: str) ->pd.DataFrame():
  """ Get the different fundamental ratios for a given ticker. """
  companies = Toolkit(
      [ticker], api_key=FINANCIAL_MODELING_PREP_API_KEY, start_date="2022-01-01"
  )
  ratios = companies.ratios.collect_all_ratios()

  print("ratios", ratios.loc['Return on Assets'])

  current_state = await ctx.get("state")

  if current_state["ticker"] == "":
    current_state["ticker"] = ticker

  if current_state["ratios"].empty:
    current_state["ratios"] = ratios

  await ctx.set("state", current_state)

  return f"Ratios extracted for {ticker}."


async def get_profitability_ratios(ctx: Context):
  """Get profitability ratios for a given ticker: ROA, ROE, Net Profit Margin and Gross Margin, and comments on these ratios given a set of threshold values."""

  current_state = await ctx.get("state")
  ratios = current_state['ratios']
  ticker = current_state['ticker']

  ROA = ratios.loc['Return on Assets']
  ROE = ratios.loc['Return on Equity']
  net_profit_margin = ratios.loc['Net Profit Margin']
  gross_margin = ratios.loc['Gross Margin']

  print("## Profitability Ratios (Assessing Earnings & Efficiency)")
  print("Return on Assets (ROA):", ROA.index[-1], ROA.iloc[-1])
  print("Return on Equity (ROE):", ROE.index[-1], ROE.iloc[-1])
  print("Net Profit Margin:", net_profit_margin.index[-1], net_profit_margin.iloc[-1])
  print("Gross Margin:", gross_margin.index[-1], gross_margin.iloc[-1])
  roa_values = [ROA.index[-1], ROA.iloc[-1]]
  roe_values = [ROE.index[-1], ROE.iloc[-1]]
  net_profit_margin_values = [net_profit_margin.index[-1], net_profit_margin.iloc[-1]]
  gross_margin_values = [gross_margin.index[-1], gross_margin.iloc[-1]]

  dico_ratios_profitability = {
      "Return on Assets (ROA)": roa_values,
      "Return on Equity (ROE)": roe_values,
      "Net Profit Margin": net_profit_margin_values,
      "Gross Margin": gross_margin_values
  }

  #Need to add conditions whether ratios, roa, roe, net profit margin, gross margin are empty or not
  if current_state["profitability_ratios"] == {}:
    current_state["profitability_ratios"] = dico_ratios_profitability

  print("current_state['profitability_ratios'] ==> after modifying state ==>", current_state["profitability_ratios"])

  thresholds_to_respect ="""
  ## Thresholds to respect for firm's financial health:
  ## Profitability Ratios (Assessing Earnings & Efficiency)
  |Ratio|	Healthy|	Moderate|	Weak|
  |-----|	-------|	--------|	----|
  |Return on Assets (ROA)	|> 5%|	2% - 5%	|< 2%|
  |Return on Equity (ROE)|	> 15%	|8% - 15%	|< 8%|
  |Net Profit Margin	|> 10%	|5% - 10%	|< 5%|
  |Gross Profit Margin	|> 40%	|20% - 40%	|< 20%|
  """

  prompt = f"""
  Analyze the financial health of the firm {ticker} based on its profitability ratios.

  ### Given:
  - **Profitability Ratios for {ticker}:**
    {dico_ratios_profitability}

  - **Thresholds for Financial Health Evaluation:**
    {thresholds_to_respect}

  ### Task:
  For each ratio, follow these steps:
  1️⃣ **Assign a score** from **1 to 10** (where **1 = very unhealthy**, **10 = very healthy**).
  2️⃣ **Provide a justification** explaining why the ratio received that score.
  3️⃣ **Give an overall insight** on the firm’s financial health, summarizing strengths and weaknesses based on the individual ratio scores.

  Ensure the analysis is **detailed, data-driven, and easy to interpret**.
  """

  resp = llm.complete(prompt)
  current_state['threshold_profitability_comments'] = resp

  print('resp from LLLM', resp)

  await ctx.set("state", current_state)

  return "Profitability ratios extracted and Comments performed: " + str(resp)


async def get_liquidity_ratios(ctx: Context):
  """Get liquidity ratios for a given ticker: Current Ratio, Quick Ratio, Debt-to-Equity Ratio, Interest Coverage Ratio and comments on these ratios given a set of threshold values."""

  current_state = await ctx.get("state")
  ratios = current_state['ratios']
  ticker = current_state['ticker']

  current_ratio = ratios.loc['Current Ratio']
  quick_ratio = ratios.loc['Quick Ratio']
  debt_to_equity_ratio = ratios.loc['Debt-to-Equity Ratio']
  interest_coverage_ratio = ratios.loc['Interest Coverage Ratio']

  print("## Profitability Ratios (Assessing Earnings & Efficiency)")
  print("Current Ratio:", current_ratio.index[-1], current_ratio.iloc[-1])
  print("Quick Ratio:", quick_ratio.index[-1], quick_ratio.iloc[-1])
  print("Debt to Equity Ratio:", debt_to_equity_ratio.index[-1], debt_to_equity_ratio.iloc[-1])
  print("Interest Coverage Ratio:", interest_coverage_ratio.index[-1], interest_coverage_ratio.iloc[-1])
  current_ratio_values = [current_ratio.index[-1], current_ratio.iloc[-1]]
  quick_ratio_values = [quick_ratio.index[-1], quick_ratio.iloc[-1]]
  debt_to_equity_ratio_values = [debt_to_equity_ratio.index[-1], debt_to_equity_ratio.iloc[-1]]
  interest_coverage_ratio_values = [interest_coverage_ratio.index[-1], interest_coverage_ratio.iloc[-1]]

  dico_ratios_liquidity = {
      "Current Ratio": current_ratio_values,
      "Quick Ratio": quick_ratio_values,
      "Debt to Equity Ratio": debt_to_equity_ratio_values,
      "Interest Coverage Ratio": interest_coverage_ratio_values
  }

  #Need to add conditions whether ratios, roa, roe, net profit margin, gross margin are empty or not
  if current_state["liquidity_ratios"] == {}:
    current_state["liquidity_ratios"] = dico_ratios_liquidity

  print("current_state['liquidity_ratios'] ==> after modifying state ==>", current_state["liquidity_ratios"])

  thresholds_to_respect ="""
  ## Thresholds to respect for firm's financial health:
  ## Liquidity & Solvency Ratios (Assessing Financial Stability)
  | Ratio	| Healthy Range	 | Warning Zone	 | Risky/Dangerous  |
  | ----	| ------------	 | -----------	 | ---------------  |
  | Current Ratio	 |  1.5 - 3.0	 |  < 1.0	 | > 3.0 (excess cash)  |
  | Quick Ratio	 | > 1.0	 | < 1.0 | 	- |
  | Debt-to-Equity (D/E) | 0.3 - 1.5	 | > 2.0	 | < 0.3 (under-leveraged) |
  | Interest Coverage  | > 3.0	| 1.5 - 3.0	 | < 1.5 (high risk) |
  """

  prompt = f"""
  Analyze the financial health of the firm {ticker} based on its profitability ratios.

  ### Given:
  - **Profitability Ratios for {ticker}:**
    {dico_ratios_liquidity}

  - **Thresholds for Financial Health Evaluation:**
    {thresholds_to_respect}

  ### Task:
  For each ratio, follow these steps:
  1️⃣ **Assign a score** from **1 to 10** (where **1 = very unhealthy**, **10 = very healthy**).
  2️⃣ **Provide a justification** explaining why the ratio received that score.
  3️⃣ **Give an overall insight** on the firm’s financial health, summarizing strengths and weaknesses based on the individual ratio scores.

  Ensure the analysis is **detailed, data-driven, and easy to interpret**.
  """

  resp = llm.complete(prompt)
  current_state['threshold_liquidity_comments'] = resp

  print('resp from LLLM', resp)

  await ctx.set("state", current_state)

  return "Liquidity ratios extracted and Comments performed: " + str(resp)


async def get_overall_comments(ctx: Context, overall_comments):
  """Get comments on diffrent type of ratios. The overall_comments are given by the SupervisedAgent based on threshold_profitability_comments and threshold_liquidity_comments."""

  current_state = await ctx.get("state")
  profitability_comments = current_state['threshold_profitability_comments']
  liquidity_comments = current_state['threshold_liquidity_comments']

  if profitability_comments is None:
    return "No profitability comments found."

  if liquidity_comments is None:
    return "No liquidity comments found."

  if ~ profitability_comments is None and ~ liquidity_comments is None:
    current_state['overall_comments'] = overall_comments

  return "Overall comments done."


## Agents

In [ ]:
from llama_index.core.agent.workflow import FunctionAgent, ReActAgent

fundamental_agent = FunctionAgent(
    name="FundamentalAgent",
    description="Get various fundamental ratios for a given ticker.",
    system_prompt=(
        "You are the fundament analyst that can extract different fundamental ratios for a given ticker. "
        "Once you have extracted the fundamental financial ratios, you should hand off control to the ProfitabilityAgent to extract the profitability ratios."
        "the ResearchAgent that can search the web for information on a given topic and record notes on the topic. "
    ),
    llm=llm,
    tools=[get_fundamentals],
    can_handoff_to=["ProfitabilityAgent"],
)

profitability_agent = FunctionAgent(
    name="ProfitabilityAgent",
    description="Collect profitability ratios for a given ticker: ROA, ROE, Net Profit Margin and Gross Margin and Comment on the results given a set of threshold values.",
    system_prompt=(
        """You are the ProfitabilityAgent that can collect profitability ratios (profitability_ratios) on a given ticker.
        You collect these ratios from the FundamentalAgent.
        Once these ratios are collected in profitability_ratios, you should comment on these ratios based on the thresholds values provided in get_profitability_ratios.
        These comments must be included in threshold_profitability_comments. At the end provide ONLY these comments included in threshold_profitability_comments. DO NOT ADD anything else.
        Once the comments are done, you should hand off control to the LiquidityAgent.
        """
    ),
    llm=llm,
    tools=[get_profitability_ratios],
    can_handoff_to=["LiquidityAgent"],
)

liquidity_agent = FunctionAgent(
    name="LiquidityAgent",
    description="Collect liquidity ratios for a given ticker: Current Ratio, Quick Ratio, Debt-to-Equity Ratio, Interest Coverage Ratio and comments on these ratios given a set of threshold values.",
    system_prompt=(
        """You are the LiquidityAgent that can collect liquidity ratios (liquidity_ratios) on a given ticker.
        You collect these ratios from the FundamentalAgent.
        Once these ratios are collected in liquidity_ratios, you should comment on these ratios based on the thresholds values provided in get_liquidity_ratios.
        These comments must be included in threshold_liquidity_comments. At the end provide ONLY these comments included in threshold_profitability_comments. DO NOT ADD anything else.
        Once the comments are done, you should hand off control to the SupervisorAgent.
        """
    ),
    llm=llm,
    tools=[get_liquidity_ratios],
    can_handoff_to=["SupervisorAgent"],
)

supervisor_agent = FunctionAgent(
    name="SupervisorAgent",
    description="Provide an overall comment based on the comments coming from the ProfitabilityAgent and LiquidityAgent.",
    system_prompt=(
        "You are an fundament analyst expert and supervisor. You collect comments coming from various agent such as ProfitabilityAgent and LiquidityAgent. "
        "Based on the results in the comments coming from the ProfitabilityAgent and LiquidityAgent, provide an overall comment on the health of the firm."
        "Justify your comment with details and data. "
    ),
    llm=llm,
    tools=[get_overall_comments],
    can_handoff_to=["FundamentalAgent"],
)


## AgentWorkflow

In [ ]:
from llama_index.core.agent.workflow import AgentWorkflow

agent_workflow = AgentWorkflow(
    agents=[fundamental_agent, profitability_agent,liquidity_agent, supervisor_agent],
    root_agent=fundamental_agent.name,
    initial_state={
        "ratios": pd.DataFrame(),
        "profitability_ratios": {},
        "liquidity_ratios": {},
        "threshold_profitability_comments": None,
        "threshold_liquidity_comments": None,
        "overall_comments": None,
        "ticker": "",
    },

)

## Calling the Mult-Agent System for a specific task

In [ ]:
from llama_index.core.agent.workflow import (
    AgentInput,
    AgentOutput,
    ToolCall,
    ToolCallResult,
    AgentStream,
)

handler = agent_workflow.run(
    user_msg=(
        "Provide the fundamental analysis of Apple and comments on the financial health of the company."
    )
)

current_agent = None
current_tool_calls = ""
async for event in handler.stream_events():
    if (
        hasattr(event, "current_agent_name")
        and event.current_agent_name != current_agent
    ):
        current_agent = event.current_agent_name
        print(f"\n{'='*50}")
        print(f"🤖 Agent: {current_agent}")
        print(f"{'='*50}\n")

    elif isinstance(event, AgentOutput):
        if event.response.content:
            print("📤 Output:", event.response.content)
        if event.tool_calls:
            print(
                "🛠️  Planning to use tools:",
                [call.tool_name for call in event.tool_calls],
            )
    elif isinstance(event, ToolCallResult):
        print(f"🔧 Tool Result ({event.tool_name}):")
        print(f"  Arguments: {event.tool_kwargs}")
        print(f"  Output: {event.tool_output}")
    elif isinstance(event, ToolCall):
        print(f"🔨 Calling Tool: {event.tool_name}")
        print(f"  With arguments: {event.tool_kwargs}")


🤖 Agent: FundamentalAgent

🛠️  Planning to use tools: ['get_fundamentals', 'handoff']


Obtaining historical data: 100%|██████████| 2/2 [00:00<00:00,  8.99it/s]


ratios 2022      NaN
2023    0.275
2024   0.2613
Freq: Y-DEC, Name: Return on Assets, dtype: float64
🔨 Calling Tool: get_fundamentals
  With arguments: {'ticker': 'AAPL'}
🔧 Tool Result (get_fundamentals):
  Arguments: {'ticker': 'AAPL'}
  Output: Ratios extracted for AAPL.
🔨 Calling Tool: handoff
  With arguments: {'to_agent': 'ProfitabilityAgent', 'reason': 'Collect profitability ratios for Apple (AAPL) and comment on the results.'}
🔧 Tool Result (handoff):
  Arguments: {'to_agent': 'ProfitabilityAgent', 'reason': 'Collect profitability ratios for Apple (AAPL) and comment on the results.'}
  Output: Agent ProfitabilityAgent is now handling the request due to the following reason: Collect profitability ratios for Apple (AAPL) and comment on the results..
Please continue with the current request.

🤖 Agent: ProfitabilityAgent

🛠️  Planning to use tools: ['get_profitability_ratios']
## Profitability Ratios (Assessing Earnings & Efficiency)
Return on Assets (ROA): 2024 0.2613
Return on Equ

## GPT-4o Answer

📤 Output: Apple Inc. (AAPL) demonstrates exceptional financial health from a profitability perspective, with all profitability ratios significantly exceeding healthy thresholds.

This indicates strong operational efficiency, effective use of assets and equity, and robust profit margins, reflecting the company's ability to generate high returns and maintain cost efficiency.

However, from a liquidity standpoint, Apple faces potential challenges. The current and quick ratios are below healthy thresholds, suggesting potential liquidity concerns in meeting short-term obligations.
Additionally, the debt-to-equity ratio indicates moderate leverage, posing some financial risk.

Despite these liquidity concerns, Apple's infinite interest coverage ratio highlights its strong ability to meet interest obligations, showcasing robust earnings.

Overall, while Apple is in a strong financial position with outstanding profitability, it should address liquidity and leverage to enhance financial stability.

# LLM: GPT-4.5-preview

## Tools

In [ ]:
from llama_index.core.workflow import Context

async def get_fundamentals(ctx: Context, ticker: str) ->pd.DataFrame():
  """ Get the different fundamental ratios for a given ticker. """
  companies = Toolkit(
      [ticker], api_key=FINANCIAL_MODELING_PREP_API_KEY, start_date="2022-01-01"
  )
  ratios = companies.ratios.collect_all_ratios()

  print("ratios", ratios.loc['Return on Assets'])

  current_state = await ctx.get("state")

  if current_state["ticker"] == "":
    current_state["ticker"] = ticker

  if current_state["ratios"].empty:
    current_state["ratios"] = ratios

  await ctx.set("state", current_state)

  return f"Ratios extracted for {ticker}."


async def get_profitability_ratios(ctx: Context):
  """Get profitability ratios for a given ticker: ROA, ROE, Net Profit Margin and Gross Margin, and comments on these ratios given a set of threshold values."""

  current_state = await ctx.get("state")
  ratios = current_state['ratios']
  ticker = current_state['ticker']

  ROA = ratios.loc['Return on Assets']
  ROE = ratios.loc['Return on Equity']
  net_profit_margin = ratios.loc['Net Profit Margin']
  gross_margin = ratios.loc['Gross Margin']

  print("## Profitability Ratios (Assessing Earnings & Efficiency)")
  print("Return on Assets (ROA):", ROA.index[-1], ROA.iloc[-1])
  print("Return on Equity (ROE):", ROE.index[-1], ROE.iloc[-1])
  print("Net Profit Margin:", net_profit_margin.index[-1], net_profit_margin.iloc[-1])
  print("Gross Margin:", gross_margin.index[-1], gross_margin.iloc[-1])
  roa_values = [ROA.index[-1], ROA.iloc[-1]]
  roe_values = [ROE.index[-1], ROE.iloc[-1]]
  net_profit_margin_values = [net_profit_margin.index[-1], net_profit_margin.iloc[-1]]
  gross_margin_values = [gross_margin.index[-1], gross_margin.iloc[-1]]

  dico_ratios_profitability = {
      "Return on Assets (ROA)": roa_values,
      "Return on Equity (ROE)": roe_values,
      "Net Profit Margin": net_profit_margin_values,
      "Gross Margin": gross_margin_values
  }

  #Need to add conditions whether ratios, roa, roe, net profit margin, gross margin are empty or not
  if current_state["profitability_ratios"] == {}:
    current_state["profitability_ratios"] = dico_ratios_profitability

  print("current_state['profitability_ratios'] ==> after modifying state ==>", current_state["profitability_ratios"])

  thresholds_to_respect ="""
  ## Thresholds to respect for firm's financial health:
  ## Profitability Ratios (Assessing Earnings & Efficiency)
  |Ratio|	Healthy|	Moderate|	Weak|
  |-----|	-------|	--------|	----|
  |Return on Assets (ROA)	|> 5%|	2% - 5%	|< 2%|
  |Return on Equity (ROE)|	> 15%	|8% - 15%	|< 8%|
  |Net Profit Margin	|> 10%	|5% - 10%	|< 5%|
  |Gross Profit Margin	|> 40%	|20% - 40%	|< 20%|
  """

  prompt = f"""
  Analyze the financial health of the firm {ticker} based on its profitability ratios.

  ### Given:
  - **Profitability Ratios for {ticker}:**
    {dico_ratios_profitability}

  - **Thresholds for Financial Health Evaluation:**
    {thresholds_to_respect}

  ### Task:
  For each ratio, follow these steps:
  1️⃣ **Assign a score** from **1 to 10** (where **1 = very unhealthy**, **10 = very healthy**).
  2️⃣ **Provide a justification** explaining why the ratio received that score.
  3️⃣ **Give an overall insight** on the firm’s financial health, summarizing strengths and weaknesses based on the individual ratio scores.

  Ensure the analysis is **detailed, data-driven, and easy to interpret**.
  """

  resp = llm_45.complete(prompt)
  current_state['threshold_profitability_comments'] = resp

  print('resp from LLLM', resp)

  await ctx.set("state", current_state)

  return "Profitability ratios extracted and Comments performed: " + str(resp)


async def get_liquidity_ratios(ctx: Context):
  """Get liquidity ratios for a given ticker: Current Ratio, Quick Ratio, Debt-to-Equity Ratio, Interest Coverage Ratio and comments on these ratios given a set of threshold values."""

  current_state = await ctx.get("state")
  ratios = current_state['ratios']
  ticker = current_state['ticker']

  current_ratio = ratios.loc['Current Ratio']
  quick_ratio = ratios.loc['Quick Ratio']
  debt_to_equity_ratio = ratios.loc['Debt-to-Equity Ratio']
  interest_coverage_ratio = ratios.loc['Interest Coverage Ratio']

  print("## Profitability Ratios (Assessing Earnings & Efficiency)")
  print("Current Ratio:", current_ratio.index[-1], current_ratio.iloc[-1])
  print("Quick Ratio:", quick_ratio.index[-1], quick_ratio.iloc[-1])
  print("Debt to Equity Ratio:", debt_to_equity_ratio.index[-1], debt_to_equity_ratio.iloc[-1])
  print("Interest Coverage Ratio:", interest_coverage_ratio.index[-1], interest_coverage_ratio.iloc[-1])
  current_ratio_values = [current_ratio.index[-1], current_ratio.iloc[-1]]
  quick_ratio_values = [quick_ratio.index[-1], quick_ratio.iloc[-1]]
  debt_to_equity_ratio_values = [debt_to_equity_ratio.index[-1], debt_to_equity_ratio.iloc[-1]]
  interest_coverage_ratio_values = [interest_coverage_ratio.index[-1], interest_coverage_ratio.iloc[-1]]

  dico_ratios_liquidity = {
      "Current Ratio": current_ratio_values,
      "Quick Ratio": quick_ratio_values,
      "Debt to Equity Ratio": debt_to_equity_ratio_values,
      "Interest Coverage Ratio": interest_coverage_ratio_values
  }

  #Need to add conditions whether ratios, roa, roe, net profit margin, gross margin are empty or not
  if current_state["liquidity_ratios"] == {}:
    current_state["liquidity_ratios"] = dico_ratios_liquidity

  print("current_state['liquidity_ratios'] ==> after modifying state ==>", current_state["liquidity_ratios"])

  thresholds_to_respect ="""
  ## Thresholds to respect for firm's financial health:
  ## Liquidity & Solvency Ratios (Assessing Financial Stability)
  | Ratio	| Healthy Range	 | Warning Zone	 | Risky/Dangerous  |
  | ----	| ------------	 | -----------	 | ---------------  |
  | Current Ratio	 |  1.5 - 3.0	 |  < 1.0	 | > 3.0 (excess cash)  |
  | Quick Ratio	 | > 1.0	 | < 1.0 | 	- |
  | Debt-to-Equity (D/E) | 0.3 - 1.5	 | > 2.0	 | < 0.3 (under-leveraged) |
  | Interest Coverage  | > 3.0	| 1.5 - 3.0	 | < 1.5 (high risk) |
  """

  prompt = f"""
  Analyze the financial health of the firm {ticker} based on its profitability ratios.

  ### Given:
  - **Profitability Ratios for {ticker}:**
    {dico_ratios_liquidity}

  - **Thresholds for Financial Health Evaluation:**
    {thresholds_to_respect}

  ### Task:
  For each ratio, follow these steps:
  1️⃣ **Assign a score** from **1 to 10** (where **1 = very unhealthy**, **10 = very healthy**).
  2️⃣ **Provide a justification** explaining why the ratio received that score.
  3️⃣ **Give an overall insight** on the firm’s financial health, summarizing strengths and weaknesses based on the individual ratio scores.

  Ensure the analysis is **detailed, data-driven, and easy to interpret**.
  """

  resp = llm_45.complete(prompt)
  current_state['threshold_liquidity_comments'] = resp

  print('resp from LLLM', resp)

  await ctx.set("state", current_state)

  return "Liquidity ratios extracted and Comments performed: " + str(resp)


async def get_overall_comments(ctx: Context, overall_comments):
  """Get comments on diffrent type of ratios. The overall_comments are given by the SupervisedAgent based on threshold_profitability_comments and threshold_liquidity_comments."""

  current_state = await ctx.get("state")
  profitability_comments = current_state['threshold_profitability_comments']
  liquidity_comments = current_state['threshold_liquidity_comments']
  # ticker = current_state['ticker']

  if profitability_comments is None:
    return "No profitability comments found."

  if liquidity_comments is None:
    return "No liquidity comments found."

  if ~ profitability_comments is None and ~ liquidity_comments is None:
    current_state['overall_comments'] = overall_comments

  return "Overall comments done."


## Agents

In [ ]:
from llama_index.core.agent.workflow import FunctionAgent, ReActAgent

fundamental_agent = FunctionAgent(
    name="FundamentalAgent",
    description="Get various fundamental ratios for a given ticker.",
    system_prompt=(
        "You are the fundament analyst that can extract different fundamental ratios for a given ticker. "
        "Once you have extracted the fundamental financial ratios, you should hand off control to the ProfitabilityAgent to extract the profitability ratios."
        "the ResearchAgent that can search the web for information on a given topic and record notes on the topic. "
    ),
    llm=llm_45,
    tools=[get_fundamentals],
    can_handoff_to=["ProfitabilityAgent"],
)

profitability_agent = FunctionAgent(
    name="ProfitabilityAgent",
    description="Collect profitability ratios for a given ticker: ROA, ROE, Net Profit Margin and Gross Margin and Comment on the results given a set of threshold values.",
    system_prompt=(
        """You are the ProfitabilityAgent that can collect profitability ratios (profitability_ratios) on a given ticker.
        You collect these ratios from the FundamentalAgent.
        Once these ratios are collected in profitability_ratios, you should comment on these ratios based on the thresholds values provided in get_profitability_ratios.
        These comments must be included in threshold_profitability_comments. At the end provide ONLY these comments included in threshold_profitability_comments. DO NOT ADD anything else.
        Once the comments are done, you should hand off control to the LiquidityAgent.
        """
    ),
    llm=llm_45,
    tools=[get_profitability_ratios],
    can_handoff_to=["LiquidityAgent"],
)

liquidity_agent = FunctionAgent(
    name="LiquidityAgent",
    description="Collect liquidity ratios for a given ticker: Current Ratio, Quick Ratio, Debt-to-Equity Ratio, Interest Coverage Ratio and comments on these ratios given a set of threshold values.",
    system_prompt=(
        """You are the LiquidityAgent that can collect liquidity ratios (liquidity_ratios) on a given ticker.
        You collect these ratios from the FundamentalAgent.
        Once these ratios are collected in liquidity_ratios, you should comment on these ratios based on the thresholds values provided in get_liquidity_ratios.
        These comments must be included in threshold_liquidity_comments. At the end provide ONLY these comments included in threshold_profitability_comments. DO NOT ADD anything else.
        Once the comments are done, you should hand off control to the SupervisorAgent.
        """
    ),
    llm=llm_45,
    tools=[get_liquidity_ratios],
    can_handoff_to=["SupervisorAgent"],
)

supervisor_agent = FunctionAgent(
    name="SupervisorAgent",
    description="Provide an overall comment based on the comments coming from the ProfitabilityAgent and LiquidityAgent.",
    system_prompt=(
        "You are an fundament analyst expert and supervisor. You collect comments coming from various agent such as ProfitabilityAgent and LiquidityAgent. "
        "Based on the results in the comments coming from the ProfitabilityAgent and LiquidityAgent, provide an overall comment on the health of the firm."
        "Justify your comment with details and data. "
    ),
    llm=llm_45,
    tools=[get_overall_comments],
    can_handoff_to=["FundamentalAgent"],
)


## AgentWorkflow

In [ ]:
from llama_index.core.agent.workflow import AgentWorkflow

agent_workflow = AgentWorkflow(
    agents=[fundamental_agent, profitability_agent,liquidity_agent, supervisor_agent],
    root_agent=fundamental_agent.name,
    initial_state={
        "ratios": pd.DataFrame(),
        "profitability_ratios": {},
        "liquidity_ratios": {},
        "threshold_profitability_comments": None,
        "threshold_liquidity_comments": None,
        "overall_comments": None,
        "ticker": "",
    },

)

## Calling the Mult-Agent System for a specific task

In [ ]:
#code from llamaindex repo
from llama_index.core.agent.workflow import (
    AgentInput,
    AgentOutput,
    ToolCall,
    ToolCallResult,
    AgentStream,
)

handler = agent_workflow.run(
    user_msg=(
        "Provide the fundamental analysis of Apple and comments on the financial health of the company."
    )
)

current_agent = None
current_tool_calls = ""
async for event in handler.stream_events():
    if (
        hasattr(event, "current_agent_name")
        and event.current_agent_name != current_agent
    ):
        current_agent = event.current_agent_name
        print(f"\n{'='*50}")
        print(f"🤖 Agent: {current_agent}")
        print(f"{'='*50}\n")
    elif isinstance(event, AgentOutput):
        if event.response.content:
            print("📤 Output:", event.response.content)
        if event.tool_calls:
            print(
                "🛠️  Planning to use tools:",
                [call.tool_name for call in event.tool_calls],
            )
    elif isinstance(event, ToolCallResult):
        print(f"🔧 Tool Result ({event.tool_name}):")
        print(f"  Arguments: {event.tool_kwargs}")
        print(f"  Output: {event.tool_output}")
    elif isinstance(event, ToolCall):
        print(f"🔨 Calling Tool: {event.tool_name}")
        print(f"  With arguments: {event.tool_kwargs}")


🤖 Agent: FundamentalAgent

🛠️  Planning to use tools: ['get_fundamentals']


Obtaining historical data: 100%|██████████| 2/2 [00:00<00:00,  9.56it/s]


ratios 2022      NaN
2023    0.275
2024   0.2613
Freq: Y-DEC, Name: Return on Assets, dtype: float64
🔨 Calling Tool: get_fundamentals
  With arguments: {'ticker': 'AAPL'}
🔧 Tool Result (get_fundamentals):
  Arguments: {'ticker': 'AAPL'}
  Output: Ratios extracted for AAPL.
🛠️  Planning to use tools: ['handoff']
🔨 Calling Tool: handoff
  With arguments: {'to_agent': 'ProfitabilityAgent', 'reason': 'Fundamental ratios for Apple have been extracted. Proceed to analyze profitability ratios and comment on financial health.'}
🔧 Tool Result (handoff):
  Arguments: {'to_agent': 'ProfitabilityAgent', 'reason': 'Fundamental ratios for Apple have been extracted. Proceed to analyze profitability ratios and comment on financial health.'}
  Output: Agent ProfitabilityAgent is now handling the request due to the following reason: Fundamental ratios for Apple have been extracted. Proceed to analyze profitability ratios and comment on financial health..
Please continue with the current request.

🤖 Agen

## GPT-4.5 Answer

## Overall Financial Health Assessment for Apple Inc. (AAPL):

### Profitability Analysis:
- **Return on Assets (ROA):** Very Healthy (26.13%)
- **Return on Equity (ROE):** Very Healthy (157.41%)
- **Net Profit Margin:** Very Healthy (23.97%)
- **Gross Profit Margin:** Very Healthy (46.21%)
- **Overall Profitability:** Apple demonstrates outstanding profitability and operational efficiency, with exceptional returns and strong margins. The company is highly effective in asset utilization, shareholder returns, and cost management.

### Liquidity Analysis:
- **Current Ratio:** Below Healthy Threshold (0.8673)
- **Quick Ratio:** Below Healthy Threshold (0.745)
- **Debt-to-Equity Ratio:** Moderately Healthy (1.8723)
- **Interest Coverage Ratio:** Extremely Healthy (Infinite)
- **Overall Liquidity:** Apple shows strong debt management and exceptional interest coverage, indicating robust financial stability. However, the current and quick ratios suggest potential short-term liquidity concerns, which should be monitored closely.

### 📌 Final Assessment:
Apple Inc. exhibits exceptional profitability and robust financial health, supported by strong operational efficiency and effective debt management. The company's profitability metrics are outstanding, reflecting strong competitive advantages and effective management practices. However, liquidity ratios indicate potential short-term vulnerabilities. Given Apple's market dominance, consistent cash flow generation, and strong brand, these liquidity concerns are likely manageable but warrant careful monitoring.

### 🏅 **Overall Financial Health Score:** 8.5/10 (Very Healthy, with minor liquidity concerns)

In conclusion, Apple remains financially strong with minor areas for improvement in liquidity management. Investors should continue to monitor liquidity ratios closely, but overall, Apple's financial position remains robust and healthy.